daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionStatsSpec.scala

```diff
@@ -0,0 +1,153 @@
+    b.fetch(create(b, byKey), byKey, None)
+
+  private[this] def lookup(b: TxBuilder) =
+    b.lookupByKey(create(b, withKey = true), true)
+
+  private[this] def rollback(b: TxBuilder) =
+    b.rollback()
+
+  private[this] def addAllNodes(b: TxBuilder, rbId: NodeId) = {
+    b.add(create(b), rbId)
+    b.add(exe(consuming = false, byKey = false)(b), rbId)
+    b.add(exe(consuming = true, byKey = false)(b), rbId)
+    b.add(exe(consuming = false, byKey = true)(b), rbId)
+    b.add(exe(consuming = true, byKey = true)(b), rbId)
+    b.add(fetch(byKey = false)(b), rbId)
+    b.add(fetch(byKey = true)(b), rbId)
+    b.add(lookup(b), rbId)
+    b.add(rollback(b), rbId)
+    ()
+  }
+
+  private[this] def repeat(n: Int, f: TxBuilder => Unit)(b: TxBuilder) =
+    (0 until n).foreach(_ => f(b))
+
+  "TransactionStats.stats" should {
+
+    val n = 3
+
+    type Getter = TransactionNodeStatistic => Int
+    val testCases = Table[TxBuilder => Node, Getter](
+      "makeNode" -> "getter",
+      (create(_), _.creates),
+      (exe(consuming = true, byKey = false), _.consumingExerciseByCids),
+      (exe(consuming = false, byKey = false), _.nonconsumingExerciseByCids),
+      (exe(consuming = true, byKey = true), _.consumingExerciseByKeys),
+      (exe(consuming = false, byKey = true), _.nonconsumingExerciseByKeys),
+      (fetch(byKey = false), _.fetcheByCids),
+      (fetch(byKey = true), _.fetchByKeys),
+      (lookup, _.lookupByKeys),
+      (rollback, _.rollbacks),
+    )
+
+    "count each type of committed nodes properly" in {
+      forEvery(testCases) { (makeNode, getter) =>
+        val builder = TxBuilder()
+
+        for (i <- 1 to n) {
+          builder.add(makeNode(builder))
+          inside(TransactionNodeStatistic.stats(builder.build())) { case (committed, rollbacked) =>
+            getter(committed) shouldBe i
+            committed.nodes shouldBe i
+            rollbacked shouldBe TransactionNodeStatistic.Empty
+          }
+        }
+      }
+    }
+
+    "count each type of rollbacked nodes properly" in {
+      forEvery(testCases) { case (makeNode, getter) =>
+        val builder = TxBuilder()
+        val rollbackId = builder.add(builder.rollback())
+
+        for (i <- 1 to n) {
+          builder.add(makeNode(builder), rollbackId)
+          inside(TransactionNodeStatistic.stats(builder.build())) { case (committed, rollbacked) =>
+            committed shouldBe TransactionNodeStatistic(0, 0, 0, 0, 0, 0, 0, 0, rollbacks = 1)
+            getter(rollbacked) shouldBe i
+            rollbacked.nodes shouldBe i
+          }
+        }
+      }
+    }
+
+    "count all committed nodes properly" in {
+      val b = TxBuilder()
+      var exeId = b.add(exe(false, false)(b))
+
+      for (i <- 1 to n) {
+        addAllNodes(b, exeId)
+        inside(TransactionNodeStatistic.stats(b.build())) { case (committed, rollbacked) =>
+          committed shouldBe TransactionNodeStatistic(i, i, 2 * i, i, i, i, i, i, i)
```

Comment:
I recommend adding a comment on the `2 * i` and the `2 * i - 1` below and maybe consider using named arguments. It’s hard to see what is differing and why.